OpenCL_C.txt

```diff
@@ -2486,95 +2739,107 @@ lp = p; // compile-time error
 
-If type `T` is qualified by the address space qualifier for address space
-`A`, then " `T` is in `A` ".
-If type `T` is in address space `A`, a pointer to `T` is also a " pointer
-into `A` " and the referenced address space of the pointer is `A`.
+lp = gp; // compile-time error
+lp = pp; // compile-time error
+lp = cp; // compile-time error
 
-A pointer to `void` in any address space shall have the same representation
-and alignment requirements as a pointer to a character type in the same
-address space.
-Similarly, pointers to differently access-qualified versions of compatible
-types shall have the same representation and alignment requirements.
-All pointers to structure types in the same address space shall have the
-same representation and alignment requirements as each other.
-All pointers to union types in the same address space shall have the same
-representation and alignment requirements as each other.
+pp = lp; // compile-time error
+pp = gp; // compile-time error
+pp = cp; // compile-time error
 
-*Clause 6.3.2.3 - Pointers, replace the first two paragraphs with the
-following paragraphs*:
+cp = lp; // compile-time error
+cp = pp; // compile-time error
+cp = gp; // compile-time error
+----------
 
-If a pointer into one address space is converted to a pointer into another
-address space, then unless the original pointer is a null pointer or the
-location referred to by the original pointer is within the second address
-space, the behavior is undefined.
-(For the original pointer to refer to a location within the second address
-space, the two address spaces must overlap).
+Example below demonstrates explicit conversions for pointers pointing to
+different address spaces.
 
-A pointer to `void` in any address space may be converted to or from a
-pointer to any incomplete or object type.
-A pointer to any incomplete or object type in some address space may be
-converted to a pointer to `void` in an enclosing address space and back
-again; the result shall compare equal to the original pointer.
+[source,c]
+----------
+// Note: these examples assume OpenCL C 2.0 or the
+// __opencl_c_generic_address_space feature support.
 
-For any qualifier _q_, a pointer to a non-_q_-qualified type may be
-converted to a pointer to the _q_-qualified version of the type (but with
-the same address-space qualifier or the generic address space); the values
-stored in the original and converted pointers shall compare equal.
+global int *gp;
+local int *lp;
+private int *pp;
+constant int *cp;
 
-*Clause 6.3.2.3 - Pointers, replace the last sentence of paragraph 4 with*:
+int *p;
+gp = (global int *)lp; // illegal to cast between named address spaces
+p = (int *)lp; // legal to cast from global to generic
+gp = (global int*)p; // legal to cast from generic to global
+----------
 
-Conversion of a null pointer to another pointer type yields a null pointer
-of that type.
-Any two null pointers whose referenced address spaces overlap shall compare
-equal.
```

Comment:
NOT EDITORIAL.

Reply:
The second sentence is part of the Embedded C spec, section 5.3:
>  Any two null pointers whose referenced address spaces overlap shall compare equal.

I don't believe OpenCL changes this.

Btw, the first sentence doesn't seem to follow the Embedded C strickly but I am trying to understand how this could be implemented since we don't have `NULL` pointer constant for different address spaces? I was trying to think of a code example of this behavior:

__private int* p1 = NULL; // no conversion, NULL is in __private address space here
__local int* p2 = NULL; // no conversion, NULL is in __local address space here
int* gen = p1; // the parser doesn't know what values p1 has, so it can only perform regular address space conversion.

Reply:
> The second sentence is part of the Embedded C spec, section 5.3:

Ah, of course.

For the first sentence, going back to C99, it says:

> Conversion  of  a  null  pointer  to  another  pointer  type  yields  a  null  pointer  of  that  type.  Any two null pointers shall compare equal.

So the only change here is w.r.t. to comparing NULL, and your quote from Embedded C already covers that.

Thanks for the clarification, no further action required, I'll resolve this.